qadence/errors/readout.py

```diff
@@ -0,0 +1,151 @@
+
+
+def corrupt(bitflip_proba: float, counters: list[Counter], n_qubits: int) -> list[Counter]:
+    def flip_bits(bitstring: str, corruption: int, n_qubits: int) -> str:
+        """Flip bits for the corruption int in bitstring."""
+        str_format = "{:0" + str(n_qubits) + "b}"
+        return str_format.format(int(bitstring, 2) ^ corruption)
+
+    # Get a tensor of random bit indices.
+    rands = where(rand(n_qubits) < bitflip_proba)[0]
+
+    if rands.numel():
+        # Corruption int.
+        corruption = torch.tensor([2**n for n in rands], dtype=torch.int64).sum().item()
+        corrupted_counters = []
+        for counter in counters:
+            corrupted_counter: Counter = Counter()
+            for bitstring, count in counter.items():
+                corrupted_bitstring = flip_bits(
+                    bitstring=bitstring, corruption=corruption, n_qubits=n_qubits
+                )
+                corrupted_counter[corrupted_bitstring] = count
+            corrupted_counters.append(corrupted_counter)
+        return corrupted_counters
+
+    return counters
+
+
+def error(
+    counters: list[Counter],
+    n_qubits: int,
+    options: dict = dict(),
+) -> list[Counter]:
+    """
+    Implements a simple uniform readout error model for position-independent bit string
+    corruption.
+
+    Args:
+        counters: Samples of bit string as Counters.
+        n_qubits: Number of shots to sample.
+        seed: Random seed value if any.
+        error_probability: Uniform error probability of wrong readout at any position
+        in the bit strings.
+        noise_distribution: Noise distribution.
+
+    Returns:
+        Samples of corrupted bit strings as list[Counter].
+    """
+
+    seed = options.get("seed")
+    error_probability = options.get("error_probability", 0.1)
+    noise_distribution = options.get("noise_distribution", WhiteNoise.UNIFORM)
+
+    # option for reproducibility
+    if seed is not None:
+        torch.manual_seed(seed)
+
+    # corrupted_bitstrings = []
+    # for counter in counters:
+    #     corrupted_bitstrings.append(
+    #         Counter(
+    #             chain(
+    #                 *[
+    #                     bs_corruption(
+    #                         bitstring=bitstring,
+    #                         n_shots=n_shots,
+    #                         error_probability=error_probability,
+    #                         noise_distribution=noise_distribution,
+    #                         n_qubits=n_qubits,
+    #                     )
+    #                     for bitstring, n_shots in counter.items()
+    #                 ]
+    #             )
+    #         )
+    #     )
+
+    # bitflip_proba = options.get("bitflip_proba")
+    # breakpoint()
+    # if bitflip_proba is None:
+    #     KeyError("Readout error protocol requires a 'bitflip_proba' option of type 'float'.")
+    return corrupt(bitflip_proba=error_probability, counters=counters, n_qubits=n_qubits)
```

Comment:
As we discussed, this is the 'original' code used for the readout mitigation, but the other version should be able to corrupt bitstrings and return correct counts.

Reply:
Tested out the commented function and it returns the same results.